### PR TITLE
[ci] Build bitstream with bazel

### PR DESCRIPTION
**Pull request by milesdai**
This PR uses Bazel to build and splice the bitstream in CI instead of the shell scripts and refactors the `azure-pipelines.yml` file.
1. Use the Bazel target instead of the `ci/scripts/build-bitstream-vivado.sh` to build the bitstreams
    - Building and splicing now happen in the same step. `chip_earlgrey_cw310_splice_rom` can now be removed
    - `chip_earlgrey_cw310` no longer depends on `cw310_sw_build` since Bazel will now build the necessary dependencies
    - The GCP bitstream upload is moved out of the splice step into its own step
2. Clean up the hyperdebug rule
    - Use our custom `bazel outquery` rule to find generated outputs
    - Remove dependency on `cw310_sw_build` step; this step can now be removed entirely
3. Update website documentation

Addresses Step 1 of #15007

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -421,18 +384,30 @@ jobs:
...
     displayName: Prepare cached bitstream
   - bash: |
-      set -e
+      set -ex
+      . util/build_consts.sh
       module load "xilinx/vivado/$(VIVADO_VERSION)"
-      ci/scripts/build-bitstream-vivado.sh top_earlgrey cw310
+      SUB_PATH=hw/top_earlgrey
+      mkdir -p $OBJ_DIR/$SUB_PATH $BIN_DIR/$SUB_PATH
+      ci/bazelisk.sh build //hw/bitstream/vivado:standard --define bitstream=vivado
```
Comment (cfrantz) [inline azure-pipelines.yml:392]: Is the `--define` required?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -421,18 +384,30 @@ jobs:
...
+      cp -rLvt $OBJ_DIR/$SUB_PATH/ \
+        $($REPO_TOP/bazelisk.sh outquery-all //hw/bitstream/vivado:fpga_cw310)
+      cp -rLvt $BIN_DIR/$SUB_PATH/ \
+        $($REPO_TOP/bazelisk.sh outquery-all //hw/bitstream/vivado:standard)
```
Comment (cfrantz) [inline azure-pipelines.yml:396]: oh, I didn't know about `-t`.  TIL.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -421,18 +384,30 @@ jobs:
...
+
+      # Rename bitstreams to be compatible with subsequent steps
+      # TODO(#13807): replace this after choosing a naming scheme
```
Comment (cfrantz) [inline azure-pipelines.yml:399]: Please speak with @a-will about the bitstream archives we're publishing to GCP.  He has some ideas about using more standard filenames and including a manifest file.

We don't need to do that work in this PR, just be aware.

**Review by cfrantz — APPROVED** [review]

File: azure-pipelines.yml [hunk 4]
```diff
@@ -399,18 +368,12 @@ jobs:
...
   condition: and(succeeded(), eq(dependencies.lint.outputs['DetermineBuildType.onlyDocChanges'], '0'), eq(dependencies.lint.outputs['DetermineBuildType.onlyDvChanges'], '0'), eq(dependencies.lint.outputs['DetermineBuildType.onlyCdcChanges'], '0'))
   pool: ci-public
   timeoutInMinutes: 180
   steps:
   - template: ci/checkout-template.yml
   - template: ci/install-package-dependencies.yml
-  - template: ci/download-artifacts-template.yml
-    parameters:
-      downloadPartialBuildBinFrom:
-        - cw310_sw_build
```
Comment (a-will) [inline azure-pipelines.yml:413]: The cached bitstream flow is now broken. We'll need to fix prepare-cached-bitstream.sh to build the test_rom and splice it into the cached bitstream.

**Review by a-will — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml (hunk 1 above)
Comment (milesdai) [reply azure-pipelines.yml:392]: Nope - removed

File: azure-pipelines.yml (hunk 3 above)
Comment (milesdai) [reply azure-pipelines.yml:399]: Got it -- #13807 is a pre-existing issue that contains the discussion on this I believe. We can revisit it to follow up on this.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -421,18 +415,30 @@ jobs:
...
+      mkdir -p $OBJ_DIR/$SUB_PATH $BIN_DIR/$SUB_PATH
+      ci/bazelisk.sh build //hw/bitstream/vivado:standard --define bitstream=vivado
+      cp -rLvt $OBJ_DIR/$SUB_PATH/ \
+        $($REPO_TOP/bazelisk.sh outquery-all //hw/bitstream/vivado:fpga_cw310)
+      cp -rLvt $BIN_DIR/$SUB_PATH/ \
+        $($REPO_TOP/bazelisk.sh outquery-all //hw/bitstream/vivado:standard)
+
+      # Rename bitstreams to be compatible with subsequent steps
+      # TODO(#13807): replace this after choosing a naming scheme
+      mv $BIN_DIR/$SUB_PATH/lowrisc_systems_chip_earlgrey_cw310_0.1.bit $BIN_DIR/$SUB_PATH/lowrisc_systems_chip_earlgrey_cw310_0.1.bit.orig
```
Comment (dmcardle) [inline azure-pipelines.yml:400]: Nit: The filename is kind of long and hard to scan. I would just omit the second occurrence, replacing the second argument with `$BIN_DIR/$SUB_PATH/`.

File: azure-pipelines.yml [hunk 6]
```diff
@@ -421,18 +415,30 @@ jobs:
...
+      mv $BIN_DIR/$SUB_PATH/fpga_cw310_rom.bit $BIN_DIR/$SUB_PATH/lowrisc_systems_chip_earlgrey_cw310_0.1.bit.splice
```
Comment (dmcardle) [inline azure-pipelines.yml:401]: Nit: The shellcheck living inside my brain says we should quote all these file paths :)

File: azure-pipelines.yml [hunk 7]
```diff
@@ -485,51 +491,47 @@ jobs:
...
   - bash: |
       set -e
-      . util/build_consts.sh
-
       module load "xilinx/vivado/$(VIVADO_VERSION)"
-
-      util/fpga/splice_rom.sh -t cw310 -T earlgrey -b PROD
-
-    displayName: Splicing bitstream with Vivado
+      ci/scripts/build-bitstream-vivado.sh top_englishbreakfast cw305
```
Comment (dmcardle) [inline azure-pipelines.yml:477]: I'm currently zoomed into the first commit — isn't this the script we were going to remove?  Or were we only planning to remove it for earlgrey?

File: azure-pipelines.yml (hunk 4 above)
Comment (dmcardle) [reply azure-pipelines.yml:413]: Ah, so the cached bitstream flow broken because artifacts from `cw310_sw_build` are no longer available? In that case, could `prepare-cached-bitstream.sh` just be replaced with `ci/bazelisk.sh build --define bitstream=gcp_splice //hw/bitstream:test_rom`?

**Review by dmcardle — COMMENTED** [review]
Partial review, just a few nits and questions. Thanks for working on this, Miles!

File: azure-pipelines.yml (hunk 4 above)
Comment (a-will) [reply azure-pipelines.yml:413]: I've not tried that rule, so I couldn't say, but that might do most of it. The original outputs of this job were a bitstream with a newly-built test_rom spliced in. The bitstream could be generated in this job or it could come from the cache.

However, the step that uses the cached bitstream ran the splice_bitstream.sh shell script, and that assumed the test_rom artifacts were already present. In addition, It looks like this job now _also_ has an output with the prod rom spliced in. The cached bitstream path will need to be synced up with this new job's flow.

Comment (milesdai) [reply azure-pipelines.yml:413]: Thanks for catching this! I'm looking into this now, and it looks like it should be possible to find the right Bazel target(s) to match this up with what was there before. I think running something like `ci/bazelisk.sh build gcp_spliced_test_rom gcp_spliced_rom gcp_spliced_rom_otp_dev` and then doing some file-moving will probably work.

File: azure-pipelines.yml (hunk 7 above)
Comment (milesdai) [reply azure-pipelines.yml:477]: We don't currently have a bazel rule for englishbreakfast. This will ultimately need to be resolved by a more comprehensive multi-top solution, but for now, we won't remove the shell scripts entirely.

File: azure-pipelines.yml (hunk 5 above)
Comment (milesdai) [reply azure-pipelines.yml:400]: Sorry, not quite understanding this -- could you clarify what you're suggesting?

Comment (dmcardle) [reply azure-pipelines.yml]: Sorry, that suggestion didn't make sense. I didn't notice that we were tacking on `.orig` to the end of the filename.

File: azure-pipelines.yml (hunk 4 above)
Comment (milesdai) [reply azure-pipelines.yml:402]: Okay I've modified the cached bitstream flow forced it to run through CI. It looks like it's passing everything. Take a look at `prepare-cached-bitstream.sh` for the modifications.

Comment (a-will) [reply azure-pipelines.yml:402]: Thanks @milesdai !

**Review by a-will — APPROVED** [review]

**Review by dmcardle — APPROVED** [review]
LGTM! (% TODOs)

I have to admit I found it hard to follow the changes to the job dependency graph. Is there a diagram somewhere? I was toying with the idea of making a script that parses out the job dependencies and pipes into GraphViz :)

**Comment by milesdai** [comment]
> LGTM! (% TODOs)
> 
> I have to admit I found it hard to follow the changes to the job dependency graph. Is there a diagram somewhere? I was toying with the idea of making a script that parses out the job dependencies and pipes into GraphViz :)

Agreed! It is very confusing which could be a good reason for making that yaml file smaller :).

I have actually made such a script already. Admittedly, it's kind of hacked together, but you can find it here: https://github.com/lowRISC/opentitan-azure-scripts.
